src/nnet3/nnet-chain-training.cc

```diff
@@ -217,17 +253,33 @@ void NnetChainTrainer::UpdateParamsWithMaxChange() {
            << "(smallest factor=" << min_scale << " on "
            << component_name_with_min_scale
            << " with max-change=" << max_change_with_min_scale <<"). ";
-    if (param_delta > nnet_config.max_param_change)
+    if (param_delta > max_param_change)
       ostr << "Global max-change factor was "
-           << nnet_config.max_param_change / param_delta
+           << max_param_change / param_delta
            << " with max-change=" << nnet_config.max_param_change << ".";
     KALDI_LOG << ostr.str();
   }
   // applies both of the max-change scalings all at once, component by component
   // and updates parameters
-  scale_factors.Scale(scale);
-  AddNnetComponents(*delta_nnet_, scale_factors, scale, nnet_);
-  ScaleNnet(nnet_config.momentum, delta_nnet_);
+  if (nnet_config.backstitch_training_scale > 0.0 && 
+      nnet_config.backstitch_training_epsilon > 0.0) {
+    const BaseFloat backstitch_ratio = nnet_config.backstitch_training_scale /
+        nnet_config.backstitch_training_epsilon;
+    BaseFloat scale_backstitch =
+        (is_backstitch_step ? -nnet_config.backstitch_training_epsilon :
+        backstitch_ratio);
+    scale_factors.Scale(scale * scale_backstitch);
+    AddNnetComponents(*delta_nnet_, scale_factors, scale * scale_backstitch,
+                      nnet_);
+    scale_backstitch = (is_backstitch_step ? (-backstitch_ratio + 1 +
+        nnet_config.backstitch_training_epsilon) / backstitch_ratio :
+        nnet_config.momentum);
```

Comment:
This code wouldn't work with momentum (you'd probably need a separate copy of the nnet for that), so better to use 0.0 instead of nnet_config.momentum, and assert that momentum is zero.   After restructuring there may be 2 paths for momentum vs backstitch, anyway.

I can't figure out whether this code is correct.  My rule when I can't easily figure out correctness is to ask for the code to be restructured, because if present me can't figure it out, then future me definitely won't be able to, not to mention present or future anyone-else.